src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2236 @@
+  // p3 : (is_up2&&'a'-'A'+char2=char1)
+  equal_exprt p1(char1, char2);
+  minus_exprt diff=minus_exprt(char_a, char_A);
+  and_exprt p2(is_upper_case_1, equal_exprt(plus_exprt(diff, char1), char2));
+  and_exprt p3(is_upper_case_2, equal_exprt(plus_exprt(diff, char2), char1));
+  return or_exprt(or_exprt(p1, p2), p3);
+}
+
+exprt string_constraint_generatort::add_axioms_for_equals_ignore_case(
+  const function_application_exprt &f)
+{
+  assert(f.type()==bool_typet() || f.type().id()==ID_c_bool);
+
+  symbol_exprt eq=fresh_boolean("equal_ignore_case");
+  typecast_exprt tc_eq(eq, f.type());
+  exprt char_a=constant_char('a');
+  exprt char_A=constant_char('A');
+  exprt char_Z=constant_char('Z');
+  string_exprt s1=add_axioms_for_string_expr(args(f, 2)[0]);
+  string_exprt s2=add_axioms_for_string_expr(args(f, 2)[1]);
+
+  // We add axioms:
+  // a1 : eq => |s1|=|s2|
+  // a2 : forall qvar, 0<=qvar<|s1|,
+  //  eq => char_equal_ignore_case(s1[qvar],s2[qvar]);
+  // a3 : !eq => |s1|!=s2 || (0 <=witness<|s1| && !char_equal_ignore_case)
+
+  implies_exprt a1(eq, s1.axiom_for_has_same_length_as(s2));
+  axioms.push_back(a1);
+
+  symbol_exprt qvar=fresh_univ_index("QA_equal_ignore_case");
+  exprt constr2=character_equals_ignore_case(
+    s1[qvar], s2[qvar], char_a, char_A, char_Z);
+  string_constraintt a2(qvar, s1.length(), eq, constr2);
+  axioms.push_back(a2);
+
+  symbol_exprt witness=fresh_exist_index("witness_unequal_ignore_case");
+  exprt zero=refined_string_typet::index_zero();
+  and_exprt bound_witness(
+    binary_relation_exprt(witness, ID_lt, s1.length()),
+    binary_relation_exprt(witness, ID_ge, zero));
+  exprt witness_eq=character_equals_ignore_case(
+    s1[witness], s2[witness], char_a, char_A, char_Z);
+  not_exprt witness_diff(witness_eq);
+  implies_exprt a3(
+    not_exprt(eq),
+    or_exprt(
+      notequal_exprt(s1.length(), s2.length()),
+      and_exprt(bound_witness, witness_diff)));
+  axioms.push_back(a3);
+
+  return tc_eq;
+}
+
+
+exprt string_constraint_generatort::add_axioms_for_length(
+  const function_application_exprt &f)
+{
+  string_exprt str=add_axioms_for_string_expr(args(f, 1)[0]);
+  return str.length();
+}
+
+exprt string_constraint_generatort::add_axioms_for_data(
+  const function_application_exprt &f)
+{
+  string_exprt str=add_axioms_for_string_expr(args(f, 3)[0]);
+  exprt tab_data=args(f, 3)[1];
+  exprt data=args(f, 3)[2];
+  symbol_exprt qvar=fresh_univ_index("QA_string_data");
+
+  // translating data[qvar]  to the correct expression
+  // which is (signed int)byte_extract_little_endian
+  // (data, (2l*qvar) + POINTER_OFFSET(byte_extract_little_endian
+  //  (tab.data, 0l, unsigned short int *)), unsigned short int)
+  mult_exprt qvar2(
+    from_integer(2, signedbv_typet(64)),
+    typecast_exprt(qvar, signedbv_typet(64)));
+  byte_extract_exprt extract(
+    ID_byte_extract_little_endian,
+    tab_data,
+    from_integer(0, signedbv_typet(64)),
+    pointer_typet(unsignedbv_typet(16)));
```

Comment:
Those hard-coded bit-widths are scary. There is a rich set of constants to take from configt.